From d9eb6bdbf54ec51eedc1ac67941c406f09f5e21b Mon Sep 17 00:00:00 2001
From: wanghaibin <wanghaibin.wang@huawei.com>
Date: Thu, 11 Jan 2024 20:43:06 +0800
Subject: [PATCH 531/960] downstream: KVM: arm64: GICv4.1: Inform the HiSilicon
 vtimer irqbypass capability

HiSilicon implementation of GICv4.1 has provided direct injection
capability of vtimer interrupts.  Inform users about the capability
by vgic_global's has_direct_vtimer, which depends on the HW-capability
from gic_kvm_info and kernel command "kvm-arm.vgic_v4_enable=1".

Signed-off-by: wanghaibin <wanghaibin.wang@huawei.com>
Signed-off-by: Zenghui Yu <yuzenghui@huawei.com>
Signed-off-by: Kunkun Jiang <jiangkunkun@huawei.com>
Signed-off-by: Dongxu Sun <sundongxu3@huawei.com>
Signed-off-by: Slim6882 <15605196882@163.com>
Signed-off-by: chenyi <chenyi211@h-partners.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 arch/arm64/kvm/vgic/vgic-v3.c |  6 ++++++
 include/kvm/arm_vgic.h        | 18 ++++++++++++++++++
 2 files changed, 24 insertions(+)

diff --git a/arch/arm64/kvm/vgic/vgic-v3.c b/arch/arm64/kvm/vgic/vgic-v3.c
index 9465d3706ab9..41e3955cee0a 100644
--- a/arch/arm64/kvm/vgic/vgic-v3.c
+++ b/arch/arm64/kvm/vgic/vgic-v3.c
@@ -653,6 +653,12 @@ int vgic_v3_probe(const struct gic_kvm_info *info)
 		kvm_info("GICv4%s support %sabled\n",
 			 kvm_vgic_global_state.has_gicv4_1 ? ".1" : "",
 			 gicv4_enable ? "en" : "dis");
+
+#ifdef CONFIG_VIRT_VTIMER_IRQ_BYPASS
+		kvm_vgic_global_state.has_direct_vtimer = info->has_vtimer && gicv4_enable;
+		if (kvm_vgic_global_state.has_direct_vtimer)
+			kvm_info("vtimer-irqbypass support enabled at GIC level\n");
+#endif
 	}
 
 	kvm_vgic_global_state.vcpu_base = 0;
diff --git a/include/kvm/arm_vgic.h b/include/kvm/arm_vgic.h
index f804210f9596..bfccd961f731 100644
--- a/include/kvm/arm_vgic.h
+++ b/include/kvm/arm_vgic.h
@@ -94,6 +94,14 @@ struct vgic_global {
 	/* Pseudo GICv3 from outer space */
 	bool			no_hw_deactivation;
 
+#ifdef CONFIG_VIRT_VTIMER_IRQ_BYPASS
+	/*
+	 * Hardware (HiSilicon implementation) has vtimer interrupt
+	 * direct injection support?
+	 */
+	bool                    has_direct_vtimer;
+#endif
+
 	/* GIC system register CPU interface */
 	struct static_key_false gicv3_cpuif;
 
@@ -423,6 +431,16 @@ void kvm_vgic_reset_mapped_irq(struct kvm_vcpu *vcpu, u32 vintid);
 
 void vgic_v3_dispatch_sgi(struct kvm_vcpu *vcpu, u64 reg, bool allow_group1);
 
+#ifdef CONFIG_VIRT_VTIMER_IRQ_BYPASS
+/**
+ * kvm_vgic_vtimer_irqbypass_support - Get the vtimer irqbypass HW capability
+ */
+static inline bool kvm_vgic_vtimer_irqbypass_support(void)
+{
+	return kvm_vgic_global_state.has_direct_vtimer;
+}
+#endif
+
 /**
  * kvm_vgic_get_max_vcpus - Get the maximum number of VCPUs allowed by HW
  *
-- 
2.43.0

